src/system_identity/http_token_retriever.rs

```diff
@@ -0,0 +1,79 @@
+use std::fmt;
+
+use http::Uri;
+
+use crate::{
+    authenticator::HttpAuthenticator,
+    http_client::HttpClient,
+    jwt::signer::{local::LocalPrivateKeySigner, JwtSignerImpl},
+    token::Token,
+    token_retriever::TokenRetrieverWithCache,
+    TokenRetriever, TokenRetrieverError,
+};
+
+use super::{client_input::AuthMethod, l1_token_retriever::L1TokenRetriever};
+
+pub enum HttpTokenRetrieverImpl<'a, C: HttpClient> {
+    ClientSecretRetriever(L1TokenRetriever<'a, C>),
+    PrivateKeyRetriever(TokenRetrieverWithCache<HttpAuthenticator<'a, C>, JwtSignerImpl>),
+}
+
+impl<C: HttpClient> fmt::Debug for HttpTokenRetrieverImpl<'_, C> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            HttpTokenRetrieverImpl::ClientSecretRetriever(retriever) => f
+                .debug_tuple("ClientSecretRetriever")
+                .field(retriever)
+                .finish(),
+            HttpTokenRetrieverImpl::PrivateKeyRetriever(retriever) => f
+                .debug_tuple("PrivateKeyRetriever")
+                .field(retriever)
+                .finish(),
+        }
+    }
+}
+
+impl<C> TokenRetriever for HttpTokenRetrieverImpl<'_, C>
+where
+    C: HttpClient,
+{
+    fn retrieve(&self) -> Result<Token, TokenRetrieverError> {
+        match self {
+            HttpTokenRetrieverImpl::ClientSecretRetriever(retriever) => retriever.retrieve(),
+            HttpTokenRetrieverImpl::PrivateKeyRetriever(retriever) => retriever.retrieve(),
+        }
+    }
+}
+
+impl<'a, C> HttpTokenRetrieverImpl<'a, C>
+where
+    C: HttpClient,
+{
+    pub fn from_auth_method(
+        http_client: &'a C,
+        auth_method: AuthMethod,
+        token_retrieval_uri: &'a Uri,
+        client_id: String,
+    ) -> Result<Self, TokenRetrieverError> {
+        match auth_method {
```

Comment:
`AuthMethod` is really coupled to the `HttpTokenRetriever` (we need to implement a way to create a TokenRetriever from the AuthMethod, I wonder if it would make sense to move `AuthMethod` to this module.

We could also consider calling this method `try_new` (or something similar). After all we use the auth method to return the TokenRetriever but we also use a set of other arguments.

Reply:
Given our earlier conversation, I think the goal would be to have the `AuthMethod` help in generating the actual `IAMClient` directly, no?

Reply:
Yeah, you'll probably kill two birds with one stone 🙂